### ISSUE TITLE
Custom wrap for set(... CACHE) 

### ISSUE THREAD

**Issue by amerlyq** [cheshirekow/cmake_format#138]
Don't really know how it can be achieved, but bunch of variables in `ProjectConfig.cmake` had much nicer alignment before -- with easy to see **type** and values/comments aligned at the end of string.
Very conscise:
```cmake
set(CMAKE_INSTALL_PREFIX "${CMAKE_BINARY_DIR}/_install"
  CACHE PATH "Directory to install produced artifacts")

set(CMAKE_TOOLCHAIN_FILE ""
  CACHE FILEPATH "")

set(BUILD_TESTING ON
  CACHE BOOL "Enable tests-related stuff")
```
Now my eyes bleed:
```cmake
set(CMAKE_INSTALL_PREFIX
    "${CMAKE_BINARY_DIR}/_install"
    CACHE PATH "Directory to install produced artifacts"
)

set(CMAKE_TOOLCHAIN_FILE
    ""
    CACHE FILEPATH ""
)

set(BUILD_TESTING
    ON
    CACHE BOOL "Enable tests-related stuff"
)
```

**Comment by amerlyq** [comment]
Somewhat related to this problem -- custom nonconventional formatting is IMO clearly superior:
```cmake
set_property(SOURCE ${src} APPEND
  PROPERTY COMPILE_DEFINITIONS "${var}=${_tag}")

set_property(
  SOURCE ${src}
  APPEND
  PROPERTY COMPILE_DEFINITIONS "${var}=${_tag}"
)
```

**Comment by cheshirekow** [comment]
Yeah, `set()` is really tricky because how it's formatted depends largely on the context of whats around. A couple of notes here:

1. You probably want `max_subgroups_hwrap=3`. e.g. this formats as:
```
set(CMAKE_TOOLCHAIN_FILE "" CACHE FILEPATH "")
```
I might actually make this the default after some investigation


2. One thing that your second comment makes me realize is that `set()` could potentially behave differently depending on whether or not we know the meaning of the variable being set. For example, in `set(CMAKE_C_COMPILER_LAUNCHER` since `CMAKE_C_COMPILER_LAUNCHER` is a buit-in variable, we know that the `set()` is storing a command and not, e.g., a list of files. That could potentially provide better context on how to format it. In fact, we can even add a config option to assign the "class" of custom variables. Something along the lines of :

```
variable_class = {
    "MY_FUNKY_SCRIPT": "COMMAND",
    "MY_SOURCES": "FILE_LIST"
}
```
Which would allow us to use "command" style rules for `MY_FUNCKY_SCRIPT` and "file-list" style rules for `MY_SOURCES` without requiring annotation in-source.

**Comment by amerlyq** [comment]
Temporarily, varclass looks like the way to go, to fix formatting through project. Downside of this approach -- maintaining list of such variables when refactoring, etc and overall too tight binding of the formatting file with project.

But it highly depends on overall CMake writing style -- heavy usage of variables or heavy usage of standard functions and verbosity, or wrapping everything into custom functions... no clear solution to fit all, but with some of them it will make experience less painful.

By the way, comparing this case -- it seems we need something like "weight" from `clang-format` to decide to wrap remaining part of list or not:
```cmake
# Short and conscise
set(CMAKE_C_COMPILER_LAUNCHER   "${ccache_exe}" CACHE FILEPATH "")

# I will wrap it -- but only when I would like to write description
set(CMAKE_C_COMPILER_LAUNCHER   "${ccache_exe}"
  CACHE FILEPATH "my dummy description")

# Thats' too much description here
set(CMAKE_C_COMPILER_LAUNCHER   "${ccache_exe}"
  CACHE FILEPATH 
    "my too long dummy description for single line")

# And now it become too long to contain on single line
set(CMAKE_C_COMPILER_LAUNCHER   
  "${ccache_exe}/path/to/different/buildroot/inside/something" 
  CACHE FILEPATH "")

# And that's what it looks like when var is list
set(CMAKE_C_COMPILER_LAUNCHER_LIST   
    "${emulator_exe}" 
    "${ccache_exe}" 
  CACHE FILEPATH "Chained list of wrappers")
```

**Comment by cheshirekow** [comment]
> maintaining list of such variables when refactoring, etc and overall too tight binding of the formatting file with project.

I think that the additional ability to tag it in-line will help here. Something like:

```
set(
  myvar # file-list
  hello.h
  hello.cc
  foobar.h
  fobar.cc)
```

> it seems we need something like "weight" from clang-format

Yes, totally agree. I already plan to incorporate this.

**Comment by amerlyq** [comment]
Eager to see how it will compare to my custom formatting proposed above.
I already feel like can predict pitfalls similar to clang-format ugliness for templated boost wrapper functions :)
